### ISSUE TITLE
Can't disable breadcrumbs without restarting.

### ISSUE THREAD

**Issue by muirdm** [emacs-lsp/lsp-mode#2525]
**Describe the bug**

I updated lsp-mode and noticed the new breadcrumbs. I wanted to disable them so I set `lsp-headerline-breadcrumb-enable` to nil and re-opened the file but the breadcrumbs still show up.

After restarting Emacs, the setting seemed to be respected.

**To Reproduce**

1. Open a file with breadcrumbs enabled at the top.
2. Set `lsp-headerline-breadcrumb-enable` to nil.
3. Close and re-open file - notice breadcrumbs still show up.

**Expected behavior**

Breadcrumbs don't show up after disabling `lsp-headerline-breadcrumb-enable`.

**Which Language Server did you use**
gopls

**OS**
Mac

**Comment by ericdallo** [comment]
Not sure it's a bug, but you can disable realtime via `M-x` `lsp-headerline-breadcrumb-mode`

**Comment by yyoncho** [comment]
@ericdallo the issue is that we attach the enable handler not buffer locally.

**Comment by ericdallo** [comment]
Oh, it makes sense @yyoncho , should I change it to buffer local the add hook or this can cause other issues?

**Comment by yyoncho** [comment]
IMHO it should be fine.

**Comment by ericdallo** [comment]
Hum, it seems it's already local

**Comment by ericdallo** [comment]
https://github.com/emacs-lsp/lsp-mode/blob/master/lsp-headerline.el#L416-L420

**Comment by yyoncho** [comment]
I haven't tested, but maybe this is problematic:

>     (add-to-list 'global-mode-string '(t (:eval lsp-modeline--code-actions-string)))

We should alter buffer-local variable.

**Comment by ericdallo** [comment]
Hum, good catch, I'll test it

**Comment by ericdallo** [comment]
Opened [this draft](https://github.com/emacs-lsp/lsp-mode/pull/2526), I tested but still, the headerline shows off when opening a new buffer with the enable variable disabled

**Comment by MaskRay** [comment]
There is an error `(void-function lsp-headerline-breadcrumb-mode)` (from `lsp--auto-configure`) if the user does not `(require lsp-headerline)` first.

**Comment by PaoloTorregroza** [comment]
Same issue in emacs 26.3 whit this configuration

```lisp

(use-package lsp-mode
  :commands (lsp lsp-deferred)
  :init
  (setq lsp-keymap-prefix "C-c l")
  :config
  (lsp-enable-which-key-integration t)
  (add-hook 'lsp-after-initialize-hook 'lsp-headerline-breadcrumb-mode)
  ;;(lsp-mode . pao/lsp-mode-setup))
)

```

Note: The breadcrumb is also hiding my tabs (I'm using centaur tabs)

**Comment by yyoncho** [comment]
@

**Comment by yyoncho** [comment]
@PaoloTorregroza 

> (I already tried to disable the breadcrumb using `lsp-headerline-breadcrumb-enable` but it gives me an error `lsp-headerline-breadcrumb-enable is not defined`)

You should do `(setq lsp-headerline-breadcrumb-enable nil)` - this cannot throw that error.

**Comment by PaoloTorregroza** [comment]
@yyoncho  True, I dont know why that gives me an error before, any way it is not working

```lisp
(use-package lsp-mode
  :commands (lsp lsp-deferred)
  :init
  (setq lsp-keymap-prefix "C-c l")
  :config
  (setq lsp-headerline-enable nil)
  (lsp-enable-which-key-integration t)
  ;;(add-hook 'lsp-after-initialize-hook 'lsp-headerline-breadcrumb-mode)
  ;;(lsp-mode . pao/lsp-mode-setup))
)

```

**Comment by yyoncho** [comment]
The variable is `lsp-headerline-breadcrumb-enable`

**Comment by PaoloTorregroza** [comment]
Ouch, weird, I will try it, sorry for annoying

**Comment by aspiers** [comment]
I just ran into this and at a glance it seems that the problem is here:

https://github.com/emacs-lsp/lsp-mode/blob/591528daf0eed278e13500e058fde96bfa347b7d/lsp-mode.el#L6975-L6981

If this was ever run when `lsp-headerline-breadcrumb-enable` was `t`, then `lsp-configure-hook` contains `lsp-headerline-breadcrumb-mode` even after `lsp-headerline-breadcrumb-enable` is later set to `nil`.  I then did:

    M-: (remove-hook 'lsp-configure-hook 'lsp-headerline-breadcrumb-mode)

and the problem is gone.  So the fix is probably to change those last two lines to:

```elisp
  (if lsp-headerline-breadcrumb-enable
    (add-hook 'lsp-configure-hook 'lsp-headerline-breadcrumb-mode))
```

and similarly for the other settings which presumably have the same problem.  I'll submit a PR for this.
